### PR TITLE
Test sending message to terminated program from program

### PR DESCRIPTION
**Pull request by SabaunT**
We don't allow sending message to terminated program from extrinsic (by user). But checking destination "activeness" is quite hard from the wasm execution level, i.e. when program sends message to some other terminated program.

Here we test this do not cause a panic and returns expected errors.

@gear-tech/dev

### CODE REVIEW

**Review by clearloop — COMMENTED** [review]
shared comment for others:

This test mainly tests sending message to terminated program using proxy program ( the constructor demo) but not extrinsic ( by user ), the frame of it is 

1. upload and init a terminated program ( make it terminated via params )
2. upload and init the constructor
3. send message to the terminated program with the constructor

File: pallets/gear/src/tests.rs [hunk 1]
```diff
@@ -13170,6 +13170,77 @@ fn test_gas_allowance_exceed_with_context() {
...
+        let mail_from_proxy = MailboxOf::<Test>::iter_key(USER_1)
+            .filter_map(|(msg, _)| (msg.source() == proxy_pid).then_some(msg))
+            .next()
+            .map(|msg| msg.payload_bytes().to_vec())
+            .expect("checked");
+        assert_eq!(
+            mail_from_proxy,
+            ReplyCode::Error(ErrorReplyReason::InactiveProgram).encode()
+        );
+        let mails_from_dead_count = MailboxOf::<Test>::iter_key(USER_1)
```
[the remark this answers is not in the document]
Comment (SabaunT) [reply pallets/gear/src/tests.rs:13237]: Actually it replies the program. The `handle_reply` of the proxy sends the auto-reply's payload to user mailbox

File: pallets/gear/src/tests.rs [hunk 2]
```diff
@@ -13170,6 +13170,77 @@ fn test_gas_allowance_exceed_with_context() {
...
+        // Also check that user hasn't received anything from the dead program.
+        let mails_from_proxy_count = MailboxOf::<Test>::iter_key(USER_1)
+            .filter_map(|(msg, _)| (msg.source() == proxy_pid).then_some(msg))
+            .count();
+        assert_eq!(mails_from_proxy_count, 1);
+        let mail_from_proxy = MailboxOf::<Test>::iter_key(USER_1)
+            .filter_map(|(msg, _)| (msg.source() == proxy_pid).then_some(msg))
+            .next()
+            .map(|msg| msg.payload_bytes().to_vec())
+            .expect("checked");
```
[the remark this answers is not in the document]
Comment (SabaunT) [reply pallets/gear/src/tests.rs]: done

File: pallets/gear/src/tests.rs [hunk 3]
```diff
@@ -13170,6 +13170,77 @@ fn test_gas_allowance_exceed_with_context() {
...
+        // "Bomb" in case after refactoring runtime we accidentally allow terminated programs to be executed.
+        let handle_dead = Calls::builder().send(user_1_bytes, b"REPLY_FROM_DEAD".to_vec());
+        let (_, pid_dead) = utils::submit_constructor_with_args(
+            // Using `USER_2` not to pollute `USER_1` mailbox to make test easier.
+            USER_2,
+            b"salt1".to_vec(),
+            Scheme::predefined(init_dead, handle_dead, Calls::default()),
+            0,
+        );
+
```
[the remark this answers is not in the document]
Comment (SabaunT) [reply pallets/gear/src/tests.rs:13194]: 1. done
2. done (all `init*` and `handle*` were simplified to `init` and `handle` as the context is obvious and they are used only once).

**Review by clearloop — APPROVED** [review]

**Review by ukint-vs — APPROVED** [review]

**Pull request merged**